pkg/cloudprovider/cloudprovider.go

```diff
@@ -145,31 +155,28 @@ func (c *CloudProvider) LivenessProbe(req *http.Request) error {
 func (c *CloudProvider) GetInstanceTypes(ctx context.Context, provisioner *v1alpha5.Provisioner) ([]*cloudprovider.InstanceType, error) {
 	var nodeTemplate *v1alpha1.AWSNodeTemplate
 	var err error
-	if provisioner.Spec.ProviderRef != nil {
-		nodeTemplate, err = c.resolveNodeTemplate(ctx, nil, &v1.ObjectReference{
-			APIVersion: provisioner.Spec.ProviderRef.APIVersion,
-			Kind:       provisioner.Spec.ProviderRef.Kind,
-			Name:       provisioner.Spec.ProviderRef.Name,
-		})
-	} else {
-		nodeTemplate, err = c.resolveNodeTemplate(ctx, provisioner.Spec.Provider.Raw, nil)
+
+	var rawProvider []byte
+	if provisioner.Spec.Provider != nil {
+		rawProvider = provisioner.Spec.Provider.Raw
 	}
+	nodeTemplate, err = c.resolveNodeTemplate(ctx, rawProvider, provisioner.Spec.ProviderRef)
```

Comment:
resolveNodeTemplate could probably just take the provisioner spec and deal with the `Raw` being nil since you have special logic in this func and below `nil` is passed directly to raw which is awkward.

Reply:
I'd agree if Machine wasn't also using this function. The parameters need to be abstract enough to work for both

Reply:
oooo right 👍